### ISSUE TITLE
Zero reason code in case of wrong password while WiFi connection (IDFGH-357)

### ISSUE THREAD

**Issue by laszlopataki** [espressif/esp-idf#2359]
## Environment

- Development Kit: ESP32-DevKitC
- Core (if using chip or module): ESP32
- IDF version : //a557e8cc
- Development Env:		Make
- Operating System:		Windows
- Power Supply:         USB

## Problem Description

Wrong reason code in SYSTEM_EVENT_STA_DISCONNECTED event when the pw is wrong.
The aim would be to detemine quick and respond to the user this case.
Now the only possiblity which I can see is to counting the number of '0' (zero) reason codes.

### Expected Behavior

Getting the correct reason codes in the event in case of the wrong password.

### Actual Behavior

The reason code many times just 0. (zero)

### Steps to reproduce

1. Turn on your PC's hotspot
2. Print the reason code in the disconnect event. (see example)
3. Config the 'scan' example with a wrong pw
4. Test it in different cases
    4.1 ESP boots and tries to connect with the wrong pw
    4.2 ESP is connected but the pw has changed after that
    
### Code to reproduce this issue (scan example + 1 row)

```cpp
/* Scan Example

   This example code is in the Public Domain (or CC0 licensed, at your option.)

   Unless required by applicable law or agreed to in writing, this
   software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
   CONDITIONS OF ANY KIND, either express or implied.
*/

/*
    This example shows how to use the All Channel Scan or Fast Scan to connect
    to a Wi-Fi network.

    In the Fast Scan mode, the scan will stop as soon as the first network matching
    the SSID is found. In this mode, an application can set threshold for the
    authentication mode and the Signal strength. Networks that do not meet the
    threshold requirements will be ignored.

    In the All Channel Scan mode, the scan will end only after all the channels
    are scanned, and connection will start with the best network. The networks
    can be sorted based on Authentication Mode or Signal Strength. The priority
    for the Authentication mode is:  WPA2 > WPA > WEP > Open
*/
#include "freertos/FreeRTOS.h"
#include "freertos/event_groups.h"
#include "esp_wifi.h"
#include "esp_log.h"
#include "esp_event_loop.h"
#include "nvs_flash.h"

/*Set the SSID and Password via "make menuconfig"*/
#define DEFAULT_SSID CONFIG_WIFI_SSID
#define DEFAULT_PWD CONFIG_WIFI_PASSWORD

#if CONFIG_WIFI_ALL_CHANNEL_SCAN
#define DEFAULT_SCAN_METHOD WIFI_ALL_CHANNEL_SCAN
#elif CONFIG_WIFI_FAST_SCAN
#define DEFAULT_SCAN_METHOD WIFI_FAST_SCAN
#else
#define DEFAULT_SCAN_METHOD WIFI_FAST_SCAN
#endif /*CONFIG_SCAN_METHOD*/

#if CONFIG_WIFI_CONNECT_AP_BY_SIGNAL
#define DEFAULT_SORT_METHOD WIFI_CONNECT_AP_BY_SIGNAL
#elif CONFIG_WIFI_CONNECT_AP_BY_SECURITY
#define DEFAULT_SORT_METHOD WIFI_CONNECT_AP_BY_SECURITY
#else
#define DEFAULT_SORT_METHOD WIFI_CONNECT_AP_BY_SIGNAL
#endif /*CONFIG_SORT_METHOD*/

#if CONFIG_FAST_SCAN_THRESHOLD
#define DEFAULT_RSSI CONFIG_FAST_SCAN_MINIMUM_SIGNAL
#if CONFIG_EXAMPLE_OPEN
#define DEFAULT_AUTHMODE WIFI_AUTH_OPEN
#elif CONFIG_EXAMPLE_WEP
#define DEFAULT_AUTHMODE WIFI_AUTH_WEP
#elif CONFIG_EXAMPLE_WPA
#define DEFAULT_AUTHMODE WIFI_AUTH_WPA_PSK
#elif CONFIG_EXAMPLE_WPA2
#define DEFAULT_AUTHMODE WIFI_AUTH_WPA2_PSK
#else
#define DEFAULT_AUTHMODE WIFI_AUTH_OPEN
#endif
#else
#define DEFAULT_RSSI -127
#define DEFAULT_AUTHMODE WIFI_AUTH_OPEN
#endif /*CONFIG_FAST_SCAN_THRESHOLD*/

static const char *TAG = "scan";

static esp_err_t event_handler(void *ctx, system_event_t *event)
{
    switch (event->event_id) {
        case SYSTEM_EVENT_STA_START:
            ESP_LOGI(TAG, "SYSTEM_EVENT_STA_START");
            ESP_ERROR_CHECK(esp_wifi_connect());
            break;
        case SYSTEM_EVENT_STA_GOT_IP:
            ESP_LOGI(TAG, "SYSTEM_EVENT_STA_GOT_IP");
            ESP_LOGI(TAG, "Got IP: %s\n",
                     ip4addr_ntoa(&event->event_info.got_ip.ip_info.ip));
            break;
        case SYSTEM_EVENT_STA_DISCONNECTED:
            ESP_LOGI(TAG, "SYSTEM_EVENT_STA_DISCONNECTED");
			printf("reason: %d\n",event->event_info.disconnected.reason);
            ESP_ERROR_CHECK(esp_wifi_connect());
            break;
        default:
            break;
    }
    return ESP_OK;
}

/* Initialize Wi-Fi as sta and set scan method */
static void wifi_scan(void)
{
    tcpip_adapter_init();
    ESP_ERROR_CHECK(esp_event_loop_init(event_handler, NULL));

    wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
    ESP_ERROR_CHECK(esp_wifi_init(&cfg));
    wifi_config_t wifi_config = {
        .sta = {
            .ssid = DEFAULT_SSID,
            .password = DEFAULT_PWD,
            .scan_method = DEFAULT_SCAN_METHOD,
            .sort_method = DEFAULT_SORT_METHOD,
            .threshold.rssi = DEFAULT_RSSI,
            .threshold.authmode = DEFAULT_AUTHMODE,
        },
    };
    ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
    ESP_ERROR_CHECK(esp_wifi_set_config(ESP_IF_WIFI_STA, &wifi_config));
    ESP_ERROR_CHECK(esp_wifi_start());
}

void app_main()
{
    // Initialize NVS
    esp_err_t ret = nvs_flash_init();
    if (ret == ESP_ERR_NVS_NO_FREE_PAGES || ret == ESP_ERR_NVS_NEW_VERSION_FOUND) {
        ESP_ERROR_CHECK(nvs_flash_erase());
        ret = nvs_flash_init();
    }
    ESP_ERROR_CHECK( ret );

    wifi_scan();
}

```

## Debug Logs
Note: Those lines which start with '>>> ' are my comments, not log message. Same for '/\*    \*/'..

```
>>>Turn on the Hotspot on PC with the wrong pw
I (66) boot: ESP-IDF v3.2-dev-708-ga557e8cc-dirty 2nd stage bootloader
I (67) boot: compile time 11:25:04
I (68) boot: Enabling RNG early entropy source...
I (83) boot: SPI Speed      : 40MHz
I (96) boot: SPI Mode       : DIO
I (108) boot: SPI Flash Size : 4MB
I (121) boot: Partition Table:
I (132) boot: ## Label            Usage          Type ST Offset   Length
I (155) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (178) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (201) boot:  2 factory          factory app      00 00 00010000 00100000
I (225) boot: End of partition table
I (238) esp_image: segment 0: paddr=0x00010020 vaddr=0x3f400020 size=0x12410 ( 74768) map
I (427) esp_image: segment 1: paddr=0x00022438 vaddr=0x3ffb0000 size=0x03354 ( 13140) load
I (461) esp_image: segment 2: paddr=0x00025794 vaddr=0x3ffb3354 size=0x00000 (     0) load
I (462) esp_image: segment 3: paddr=0x0002579c vaddr=0x40080000 size=0x00400 (  1024) load
I (485) esp_image: segment 4: paddr=0x00025ba4 vaddr=0x40080400 size=0x0a46c ( 42092) load
I (618) esp_image: segment 5: paddr=0x00030018 vaddr=0x400d0018 size=0x63334 (406324) map
I (1495) esp_image: segment 6: paddr=0x00093354 vaddr=0x4008a86c size=0x05f90 ( 24464) load
I (1560) esp_image: segment 7: paddr=0x000992ec vaddr=0x400c0000 size=0x00000 (     0) load
I (1561) esp_image: segment 8: paddr=0x000992f4 vaddr=0x50000000 size=0x00000 (     0) load
I (1642) boot: Loaded app from partition at offset 0x10000
I (1642) boot: Disabling RNG early entropy source...
I (1645) cpu_start: Pro cpu up.
I (1655) cpu_start: Starting app cpu, entry point is 0x40080fe8
I (1) cpu_start: App cpu up.
I (1688) heap_init: Initializing. RAM available for dynamic allocation:
I (1709) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (1728) heap_init: At 3FFB93D0 len 00026C30 (155 KiB): DRAM
I (1747) heap_init: At 3FFE0440 len 00003BC0 (14 KiB): D/IRAM
I (1767) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (1787) heap_init: At 400907FC len 0000F804 (62 KiB): IRAM
I (1807) cpu_start: Pro cpu start user code
I (605) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
I (770) wifi: wifi driver task: 3ffc0784, prio:23, stack:3584, core=0
I (770) wifi: wifi firmware version: ece97c4
I (770) wifi: config NVS flash: enabled
I (770) wifi: config nano formating: disabled
I (780) system_api: Base MAC address is not set, read default base MAC address from BLK0 of EFUSE
I (790) system_api: Base MAC address is not set, read default base MAC address from BLK0 of EFUSE
I (820) wifi: Init dynamic tx buffer num: 32
I (820) wifi: Init data frame dynamic rx buffer num: 32
I (820) wifi: Init management frame dynamic rx buffer num: 32
I (820) wifi: Init static rx buffer size: 1600
I (830) wifi: Init static rx buffer num: 10
I (830) wifi: Init dynamic rx buffer num: 32
I (930) phy: phy_version: 3960, 5211945, Jul 18 2018, 10:40:07, 0, 0
I (930) wifi: mode : sta (30:ae:a4:3a:ca:e4)
I (940) scan: SYSTEM_EVENT_STA_START
I (1660) wifi: n:6 0, o:1 0, ap:255 255, sta:6 0, prof:1
I (2640) wifi: state: init -> auth (b0)
I (2660) wifi: state: auth -> assoc (0)
I (2670) wifi: state: assoc -> run (10)
I (9180) wifi: state: run -> init (c0)
I (9180) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (9180) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 0 /* Should be one of these: WIFI_REASON_AUTH_FAIL || WIFI_REASON_AUTH_LEAVE || WIFI_REASON_AUTH_EXPIRE */
I (9300) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (9300) wifi: state: init -> auth (b0)
I (9310) wifi: state: auth -> assoc (0)
I (9310) wifi: state: assoc -> run (10)
I (15820) wifi: state: run -> init (c0)
I (15820) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (15820) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 0
>>>Turn off Hotspot 
I (15940) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (15940) wifi: state: init -> auth (b0)
I (15950) wifi: state: auth -> assoc (0)
I (15950) wifi: state: assoc -> run (10)
I (21420) wifi: state: run -> init (8c0)
I (21420) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (21420) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 8 /* This is correct. WIFI_REASON_ASSOC_LEAVE*/
I (21550) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (21550) wifi: state: init -> auth (b0)
I (22550) wifi: state: auth -> init (200)
I (22550) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (22550) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 2 /* This is correct. WIFI_REASON_AUTH_EXPIRE*/
I (24960) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201 /* This is correct. WIFI_REASON_NO_AP_FOUND */
I (27370) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201
I (29790) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201
I (32200) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201
I (34610) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201
>>>Turn on Hotspot
I (34730) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (34730) wifi: state: init -> auth (b0)
I (34740) wifi: state: auth -> assoc (0)
I (34740) wifi: state: assoc -> run (10)
I (41250) wifi: state: run -> init (c0)
I (41250) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (41250) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 0 /* Should be one of these: WIFI_REASON_AUTH_FAIL || WIFI_REASON_AUTH_LEAVE || WIFI_REASON_AUTH_EXPIRE || WIFI_REASON_ASSOC_LEAVE */
I (41370) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (41370) wifi: state: init -> auth (b0)
I (41370) wifi: state: auth -> assoc (0)
I (41380) wifi: state: assoc -> run (10)
I (47890) wifi: state: run -> init (c0)
I (47890) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (47890) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 0
I (48010) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (48010) wifi: state: init -> auth (b0)
I (48020) wifi: state: auth -> assoc (0)
I (48020) wifi: state: assoc -> run (10)
I (54530) wifi: state: run -> init (c0)
I (54530) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (54530) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 0
>>>Turn off Hotspot
I (54650) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (54650) wifi: state: init -> auth (b0)
I (54660) wifi: state: auth -> assoc (0)
I (54660) wifi: state: assoc -> run (10)
I (54710) wifi: state: run -> init (8c0)
I (54720) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (54720) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 8
I (54840) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (54840) wifi: state: init -> auth (b0)
I (55840) wifi: state: auth -> init (200)
I (55840) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (55840) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 2
I (58260) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201
>>> Turn on Hotspot with the correct password
I (58380) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (58380) wifi: state: init -> auth (b0)
I (58380) wifi: state: auth -> assoc (0)
I (58390) wifi: state: assoc -> run (10)
I (58400) wifi: connected with Wifine, channel 6
I (58410) wifi: pm start, type: 1

I (62770) event: sta ip: 192.168.137.4, mask: 255.255.255.0, gw: 192.168.137.1
I (62770) scan: SYSTEM_EVENT_STA_GOT_IP
I (62770) scan: Got IP: 192.168.137.4

I (71800) wifi: state: run -> init (c0)
I (71800) wifi: pm stop, total sleep time: 11522719 us / 13386326 us
>>> Change the password to a wrong one without turn off on windows UI
I (71800) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (71800) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 0
I (71930) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (71930) wifi: state: init -> auth (b0)
I (72930) wifi: state: auth -> init (200)
I (72930) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (72930) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 2 /* OK: WIFI_REASON_AUTH_EXPIRE */
I (75340) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201 /* OK: WIFI_REASON_NO_AP_FOUND */
I (75470) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (75470) wifi: state: init -> auth (b0)
I (75470) wifi: state: auth -> assoc (0)
I (75480) wifi: state: assoc -> run (10)
I (81990) wifi: state: run -> init (c0)
I (81990) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (81990) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 0  /* Should be one of these: WIFI_REASON_AUTH_FAIL || WIFI_REASON_AUTH_LEAVE || WIFI_REASON_AUTH_EXPIRE || WIFI_REASON_ASSOC_LEAVE */
I (82110) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (82110) wifi: state: init -> auth (b0)
I (82120) wifi: state: auth -> assoc (0)
I (82120) wifi: state: assoc -> run (10)
I (88630) wifi: state: run -> init (c0)
I (88630) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (88630) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 0
>>> Turn off Hotspot
I (88750) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (88750) wifi: state: init -> auth (b0)
I (88760) wifi: state: auth -> assoc (0)
I (88760) wifi: state: assoc -> run (10)
I (92910) wifi: state: run -> init (8c0)
I (92910) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (92920) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 8 /* OK */
I (93040) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (93040) wifi: state: init -> auth (b0)
I (94040) wifi: state: auth -> init (200)
I (94040) wifi: n:6 0, o:6 0, ap:255 255, sta:6 0, prof:1
I (94040) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 2 /* OK */
I (96460) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201 /* OK */
I (98870) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201
I (101280) scan: SYSTEM_EVENT_STA_DISCONNECTED
reason: 201

```

**Comment by liuzfesp** [comment]
@laszlopataki this is our known issue because of some historic reasons. We will add the 4-handshake timeout reason if the password is wrong or if 4-handshake fails in future release.

**Comment by laszlopataki** [comment]
Hi @liuzfesp,

Thanks for the response.
Please let me know if you have any known release date or version number.

Regards

**Comment by liuzfesp** [comment]
Hi @laszlopataki, I hope we can complete it in IDF v3.1/3.2 bugfix release and v4.0, I can't give you the exact date and it should be done in October or November.

**Comment by zhangyanjiaoesp** [comment]
@laszlopataki I have tested the scan example as what you said. But I didn't reproduce your problem, I only get the reason 204 or 201. 
I find that when the reason is 0, the state change of the wifi is `wifi: state: run -> init (c0)` in your log, Then I check the wifi code, I find when the value is c0, it means that sta receive a deauth packet, so please check whether your AP has send deauth packet to station.

**Comment by enricop** [comment]
Hi,

in our project we have to alert the user if the Wifi credentials are wrong.
We are working with the `release/v4.0` branch.

When connecting to the right SSID with the wrong password we are receiving `REASON 201` (`WIFI_REASON_NO_AP_FOUND`) on the `WIFI_EVENT_STA_DISCONNECTED` event.

Sholdn't it be `WIFI_REASON_AUTH_FAIL` ?
Can you fix this?

Thanks for your support

**Comment by zhangyanjiaoesp** [comment]
hi @enricop,
 I have tested on the least release/v4.0 branch. 
When the AP is open, but the STA uses a password to connect the AP, the disconnect event reason is `WIFI_REASON_NO_AP_FOUND(201)`. 
When the AP is encrypted, and the STA uses a wrong password to connect the AP, the disconnect event reason is `WIFI_REASON_4WAY_HANDSHAKE_TIMEOUT(15)`.

**Comment by Alvin1Zhang** [comment]
Thanks for reporting and sharing updates. Feel free to reopen if the issue still happens, or please help file a new ticket for new issue, thanks.
